### PR TITLE
Add json file with more detailed explanations of each parameter

### PR DESCRIPTION
**Pull request by GenevieveBuckley**
This pull request belongs to the larger effort to solve briefcase [issue #392](https://github.com/beeware/briefcase/issues/392), see the [PR here](https://github.com/beeware/briefcase/pull/414)

I felt like it made more sense to store a dictionary of explanation for each of the cookiecutter parameters in the same place as those cookiecutter parameters. In my mind, this makes it more likely to be updated if needed (eg: if the template is ever changed). Feel free to disagree, it's easy enough to store this info as a dict in the main briefcase repo if you prefer.

Note that the presence of this file is optional, so you can still point to another template without one & get a relatively sensible result (just one without the nice extra explanations).

Paired with https://github.com/beeware/briefcase/pull/414

### CODE REVIEW

**Comment by GenevieveBuckley** [comment]
CI build is failing because tox can't find a python 3.5 environment, I think this is unrelated.

File: cookiecutter_context.json [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        "This is the name that will be displayed to humans whenever the ",
+        "name of the application is displayed. It can have spaces and ",
+        "punctuation if you like, and any capitalization will be ",
+        "used as you type it. "
+    ],
+    "app_name": [
+        "Next, we need a name that can serve as a machine-readable ",
+        "Python package name for your application. ",
+        "This name must be PEP508-compliant - that means the name ",
+        "may only contain letters, numbers, hypehns and underscores; ",
```
Comment (freakboy3742) [inline cookiecutter_context.json:13]: typo: hyphens.

File: cookiecutter_context.json [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        "App stores need to protect against having multiple applications ",
+        "with the same name; the bundle identifier is the namespace they use ",
+        "to identify applications that come from you. ",
+        "The bundle identifier is usually the domain name of your company or ",
+        "project, in reverse order.\n",
+        "For example, if you are writing an application for Example Corp, ",
+        "whose website is example.com, your bundle would be ``com.example``. ",
+        "The bundle will be combined with your application's ",
+        "machine readable name to form a complete application identifier ",
+        "(e.g., com.example.helloworld. "
```
Comment (freakboy3742) [inline cookiecutter_context.json:47]: Missing closing bracket)

**Comment by GenevieveBuckley** [comment]
Thanks!

**Re: ragged lines**
I generally try to break lines around semantically important parts of the sentence. I uh, didn't realise that was a "thing" I did until you mentioned it just now. I think it's because I have an easier time reading the chunked sentences that way, but it's certainly easy to change if you dislike it.

If you'd like me to change this, would you let me know what your preferred line break policy is? Eg: after the last whole word plus a space, that still leaves us under the 80 character limit?

**Re: contextual examples**
I chose to omit these because:
1. They already appear in the next line printed as the cookiecutter suggested default anyway, and
2. I wanted these explanations to be more cleanly separated from the briefcase code. If we want to turn these into function strings in briefcase, then briefcase either needs to know what variables should go into the function string, or we put the variable names in the strings here & it's brittle if variable names are ever changed in a totally separate repository. I don't like it, it feels ugly.

tl:dr - I think we get the same amount of usefulness without extra ugliness by not having them in the info strings.

**Comment by freakboy3742** [comment]
> Thanks!
> 
> **Re: ragged lines**
> I generally try to break lines around semantically important parts of the sentence. I uh, didn't realise that was a "thing" I did until you mentioned it just now. I think it's because I have an easier time reading the chunked sentences that way, but it's certainly easy to change if you dislike it.
>
> If you'd like me to change this, would you let me know what your preferred line break policy is? Eg: after the last whole word plus a space, that still leaves us under the 80 character limit?

I'm completely on board with chunking sentences into consumable pieces to improve readability. 

However, if we're going to do that, I'd like to double down on the idea and use full paragraph breaks (i.e., blank lines) and bullet points, not just mid-line breaks. Taking the `app_name` prompt as an example, compare:

     Next, we need a name that can serve as a machine-readable 
     Python package name for your application. 
     This name must be PEP508-compliant - that means the name 
     may only contain letters, numbers, hyphens and underscores; 
     it can't contain spaces or punctuation, 
     and it can't start with a hyphen or underscore. 

to

     Next, we need a name that can serve as a machine-readable Python package name 
     for your application. 

     This name must be PEP508-compliant. This means the name: 
       * may only contain letters, numbers, hyphens and underscores; 
       * can't contain spaces or punctuation; and
       * can't start with a hyphen or underscore. 
 
As for line length; the 80 char limit is really driven by "what will fit on a default 80x25 console". Applying Bringhurst "Elements of Typographic style" logic, [a 66 character limit](http://webtypography.net/2.1.2) would possibly make more sense. That would be 66 visual characters from the left margin, not including any trailing space (which are functionally non-visible).

> **Re: contextual examples**
> tl:dr - I think we get the same amount of usefulness without extra ugliness by not having them in the info strings.

I'll agree it will be at the very least difficult to communicate to template designers; possibly complex to implement; and I accept that the fact the prompt contains the "default" name will displayed removes most of the need to include a computed value in the explanatory note.

The one place where this argument falls down a little is the bundle ID, because the name "helloworld" comes out of nowhere (whereas previously it was contextually sensitive). We might be able to fix this by reworking the prompt text in that case.

File: cookiecutter_context.json [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        "First, we need a formal name for your application. ",
+        "This is the name that will be displayed to humans whenever the ",
+        "name of the application is displayed. It can have spaces and ",
+        "punctuation if you like, and any capitalization will be ",
+        "used as you type it. "
+    ],
+    "app_name": [
+        "Next, we need a name that can serve as a machine-readable ",
+        "Python package name for your application. ",
+        "\n",
```
Comment (freakboy3742) [inline cookiecutter_context.json:12]: Does this need the `\n`? Won't this render as a double newline?

File: cookiecutter_context.json [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        "This name must be PEP508-compliant. This means the name: ",
+        " * may only contain letters, numbers, hyphens and underscores; ",
+        " * can't contain spaces or punctuation; and ",
+        " * can't start with a hyphen or underscore. "
+    ],
+    "project_name": [
+        "Briefcase can manage projects that contain multiple applications, ",
+        "so we need a Project name. ",
+        "If you're only planning to have one application in this project, ",
```
Comment (freakboy3742) [inline cookiecutter_context.json:21]: Do we need a paragraph spacer blank line here? Or should these two sentences be on the same line?

File: cookiecutter_context.json [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        "or a generic contact address you set up specifically for this ",
+        "application."
+    ],
+    "bundle": [
+        "Now we need a bundle identifier for your application. ",
+        "App stores need to protect against having multiple applications ",
+        "with the same name; the bundle identifier is the namespace they use ",
+        "to identify applications that come from you. ",
+        "The bundle identifier is usually the domain name of your company or ",
+        "project, in reverse order.\n",
```
Comment (freakboy3742) [inline cookiecutter_context.json:43]: I think we should pick a style and stick with it - either use standalone `""` or `\n` for blank lines, but not both. Personally, I lean to "", as it renders most like the final output, but I could be convinced otherwise.

File: cookiecutter_context.json (hunk 3 above)
Comment (GenevieveBuckley) [reply cookiecutter_context.json:12]: No, there's no new line unless you explictly specify one with `\n`.
I've used `"".join(explanation[key])` to join these into a single string.

**Comment by GenevieveBuckley** [comment]
I think I haven't communicated clearly how newlines work here. I'm joining all the strings from the lists with `"".join(explanation[key])`. That means:
* There's no new line in the output string, unless we specify one with `\n`
* The choice of ragged/even lines in the code here only affects what you see here when you edit the code, not what the user sees when they run `briefcase new`. 

I did this because I wanted to reproduce the same output as we had before I made changes, and this seemed like the simplest way given the constraints of json.

File: cookiecutter_context.json (hunk 4 above)
Comment (GenevieveBuckley) [reply cookiecutter_context.json:21]: These two sentences do display on the same line in the output when you run `briefcase new` (although if you want we can add a `\n` character to add a line break).

File: cookiecutter_context.json (hunk 5 above)
Comment (GenevieveBuckley) [reply cookiecutter_context.json:43]: "" will not render as a line break, whereas `\n` will. I think your suggestion here is that I should put the `\n` on a separate line by itself, rather than trailing the text here?

Comment (freakboy3742) [reply cookiecutter_context.json:43]: Ok - Looks like I misunderstood how this text is being rendered.

I'm suggesting that the text in the template should be rendered with 66-char string lengths, and then be joined with `"\n"`; that way, the template shows you *exactly* what is going to be rendered to the end user, including line breaks. That also means that `""` on a line by itself appears like a paragraph break in text, *and* will be displayed as such to the end user. 

Joining with `''` means these lines are going to render as exceedingly long paragraphs, which isn't great for readability.

**Comment by GenevieveBuckley** [comment]
Update: we are now joining the explanation strings with "\n" (in the corresponding briefcase PR), and keeping as close to even string lengths here as possible.

**Comment by freakboy3742** [comment]
Closing following the comments on https://github.com/beeware/briefcase/pull/414#issuecomment-1088251728
